src/input/input_python.rs

```diff
@@ -271,15 +273,35 @@ impl<'a> Input<'a> for PyAny {
         }
     }
 
+    #[cfg(not(PyPy))]
     fn lax_list(&'a self) -> ValResult<GenericListLike<'a>> {
         if let Ok(list) = self.cast_as::<PyList>() {
             Ok(list.into())
         } else if let Ok(tuple) = self.cast_as::<PyTuple>() {
             Ok(tuple.into())
-        } else if let Ok(set) = self.cast_as::<PySet>() {
-            Ok(set.into())
-        } else if let Ok(frozen_set) = self.cast_as::<PyFrozenSet>() {
-            Ok(frozen_set.into())
+        } else if let Ok(iterator) = self.cast_as::<PyIterator>() {
+            let tuple = PyTuple::new(self.py(), iterator.iter()?.flatten().collect::<Vec<_>>());
+            Ok(tuple.into())
+        } else if let Ok(dict_keys) = self.cast_as::<PyDictKeys>() {
+            let tuple = PyTuple::new(self.py(), dict_keys.iter()?.flatten().collect::<Vec<_>>());
+            Ok(tuple.into())
+        } else if let Ok(dict_values) = self.cast_as::<PyDictValues>() {
+            let tuple = PyTuple::new(self.py(), dict_values.iter()?.flatten().collect::<Vec<_>>());
+            Ok(tuple.into())
```

Comment:
Looks like we have this lump of code in multiple cases, could you move it into a macro (or parameterised function?) and reuse?